quadra/utils/model_manager.py

```diff
@@ -0,0 +1,292 @@
+import getpass
+import os
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Dict, Optional
+
+from aigo_research.utils.utils import get_logger
+
+log = get_logger(__name__)
+
+try:
+    import mlflow  # noqa
+    from mlflow.entities import Run  # noqa
+    from mlflow.entities.model_registry import ModelVersion  # noqa
+    from mlflow.exceptions import RestException  # noqa
+    from mlflow.tracking import MlflowClient  # noqa
+
+    MLFLOW_AVAILABLE = True
+except ImportError:
+    MLFLOW_AVAILABLE = False
+
+
+VERSION_MD_TEMPLATE = "## **Version {}**\n"
+DESCRIPTION_MD_TEMPLATE = "### Description: \n{}\n"
+
+
+class AbstractModelManager(ABC):
+    """Abstract class for model managers."""
+
+    @abstractmethod
+    def register_model(self, model_location: str, model_name: str, tags: Optional[Dict[str, Any]]) -> Any:
+        """Register a model in the model registry."""
+
+    @abstractmethod
+    def get_latest_version(self, model_name: str) -> Any:
+        """Get the latest version of a model for all the possible stages or filtered by stage."""
+
+    @abstractmethod
+    def transistion_model(self, model_name: str, version: int, stage: str) -> Any:
+        """Transition the model with the given version to a new stage."""
+
+    @abstractmethod
+    def delete_model(self, model_name: str, version: int) -> None:
+        """Delete a model with the given version."""
+
+    @abstractmethod
+    def register_best_model(
+        self,
+        experiment_name: str,
+        metric: str,
+        model_name: str,
+        tags: Optional[Dict[str, Any]] = None,
+        description: str = "",
+        mode: str = "max",
+        model_path: str = "model",
+    ) -> Any:
+        """Register the best model from an experiment."""
+
+
+class MlflowModelManager(AbstractModelManager):
+    """Model manager for Mlflow."""
+
+    def __init__(self):
+        if not MLFLOW_AVAILABLE:
+            raise ImportError("Mlflow is not available, please install it with pip install mlflow")
+
+        if os.getenv("MLFLOW_TRACKING_URI") is None:
+            raise ValueError("MLFLOW_TRACKING_URI environment variable is not set")
+
+        mlflow.set_tracking_uri(os.getenv("MLFLOW_TRACKING_URI"))
+        self.client = MlflowClient()
+
+    def register_model(
+        self, model_location: str, model_name: str, tags: Optional[Dict[str, Any]] = None, description: str = ""
```

Comment:
Maybe the description should be mandatory instead of having the default value of the empty string. If we want it to be optional, we can set it to be `None`

Reply:
I'll enforce it to be mandatory even tough I'm not a huge fan, but it probably makes more senso to always require a description